dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java

```diff
@@ -59,7 +64,28 @@
 
     public MergeObject generateMergeObject( TrackedEntityInstance original, TrackedEntityInstance duplicate )
     {
-        return null;
+        if ( !duplicate.getTrackedEntityType().equals( original.getTrackedEntityType() ) )
+        {
+            throw new PotentialDuplicateForbiddenException(
+                "Potentical Duplicate does not have the same tracked entity type as the original" );
+        }
+
+        List<String> attributes = getMergeableAttributes( original, duplicate );
+
+        List<String> relationships = getMergeableRelationships( original, duplicate );
+
+        MergeObject mergeObject = MergeObject.builder()
+            .trackedEntityAttributes( attributes )
+            .relationships( relationships )
+            .build();
+
+        if ( mergeObject.getTrackedEntityAttributes().isEmpty() && mergeObject.getRelationships().isEmpty() )
+        {
+            throw new PotentialDuplicateNullException(
+                "Potentical Duplicate is null hence nothing to merge" );
```

Comment:
Does it make sense to throw an error if there is nothing to merge?
Lets say we have to identical TEIs, we still want the duplicate to be removed and the potentialDuplicate to be marked as MERGED

Reply:
not sure. what shall we do when the item to be merged has nothing to be merged?

Reply:
If the mergeObject is empty, it implies we keep the original and delete the duplicate :) No error needed

Reply:
empty mergeObject is now returned